src/allmydata/storage_client.py

```diff
@@ -254,6 +257,117 @@ def get_longname(self):
     def get_nickname(self):
         return "?"
 
+class _AnonymousStorage(object):
+    """
+    Abstraction for connecting to an anonymous storage server.
+    """
+    @classmethod
+    def from_announcement(cls, server_id, ann):
+        """
+        Create an instance from an announcement like::
+
+            {"anonymous-storage-FURL": "pb://...@...",
+             "permutation-seed-base32": "...",
+             "nickname": "...",
+            }
+
+        *nickname* is optional.
+        """
+        self = cls()
+        self._furl = str(ann["anonymous-storage-FURL"])
+        m = re.match(r'pb://(\w+)@', self._furl)
+        assert m, self._furl
+        tubid_s = m.group(1).lower()
+        self._tubid = base32.a2b(tubid_s)
+        if "permutation-seed-base32" in ann:
+            ps = base32.a2b(str(ann["permutation-seed-base32"]))
+        elif re.search(r'^v0-[0-9a-zA-Z]{52}$', server_id):
+            ps = base32.a2b(server_id[3:])
+        else:
+            log.msg("unable to parse serverid '%(server_id)s as pubkey, "
+                    "hashing it to get permutation-seed, "
+                    "may not converge with other clients",
+                    server_id=server_id,
+                    facility="tahoe.storage_broker",
+                    level=log.UNUSUAL, umid="qu86tw")
+            ps = hashlib.sha256(server_id).digest()
+        self._permutation_seed = ps
+
+        assert server_id
+        self._long_description = server_id
+        if server_id.startswith("v0-"):
+            # remove v0- prefix from abbreviated name
+            self._short_description = server_id[3:3+8]
+        else:
+            self._short_description = server_id[:8]
+        self._nickname = ann.get("nickname", "")
+        return self
+
```

Comment:
We're already using `attrs` right? This might be more-clear using attrs, or at least with a "data"-style `__init__` method that all these things are passed to (i.e. rather than having to hunt for `self.* =` to see what's assigned where)?